### ISSUE TITLE
Markers "jiggle" when panning in 1.0.0

### ISSUE THREAD

**Issue by gpolyn** [google-map-react/google-map-react#575]
As can be observed in `.mov` file linked below, markers "oscillate" or "jiggle" when the map is panned.

&nbsp;&nbsp;&nbsp;~~https://drive.google.com/open?id=1fYgvhW5Pug_XLg-adXRhWWR95QUrfOA6~~

&nbsp;&nbsp;&nbsp;https://drive.google.com/open?id=14WHc7N3HeH7JvpBKEeCl6TB2RCVEnuY1

The demo can be found at the `markers_jiggle` branch at https://github.com/gpolyn/nextjs-google-map-react, done in Next.js.

~~Can't~~ Rather not use `1.0.0` like this (at least not with Next.js.)

**Comment by itsmichaeldiego** [comment]
@gpolyn I cannot see your video: https://cl.ly/0Y2r3J1k2j1O

Why do you say you can't use `1.0.0` with NextJS? The panning of the map does not have anything to do with server-side rendering, or maybe I am missing something here...

**Comment by gpolyn** [comment]
@itsmichaeldiego ...should work now, if you want to retry.

Here's another: https://drive.google.com/open?id=14WHc7N3HeH7JvpBKEeCl6TB2RCVEnuY1

Might be useful for someone to confirm that nothing like what I've linked is found in a straight `react` use of `1.0.0`. I can do that, but but there was no jiggle in `0.0.29` in other Next work I've done, plus the fact I selected `google-map-react` as an SSR-ready tool.

**Comment by gpolyn** [comment]
...markers nice and static while panning in `0.34.0` + next.js.

&nbsp;&nbsp;&nbsp;&nbsp;https://drive.google.com/open?id=1odRsPGDbalNPYj3-OSqEZTfBo0OGrM2a

(https://github.com/gpolyn/nextjs-google-map-react/tree/markers_dont_jiggle.)

**Comment by gpolyn** [comment]
While the foregoing might be useful to others using NextJS, I see that the effect is not present at `1.0.0` https://github.com/google-map-react/google-map-react/tree/master/develop, so I'll close.

**Comment by itsmichaeldiego** [comment]
@gpolyn I would like to take a look at this, as I used NextJS without a problem in the past and I use it often, also a lot of users are doing SSR so I am interested on fixing it, let me know if I can be of any help to fix that jiggle you're mentioning. 

@stephenfarrar If you're able to see the videos @gpolyn is sharing, would you mind taking a look?

@gpolyn As I am saying, I wasn't able to see the videos, I don't know if its because of the country I am in, if you could upload them somewhere else besides drive would be awesome. Thanks in advance.

**Comment by gpolyn** [comment]
@itsmichaeldiego maybe this link will work for you: https://gfycat.com/IllegalWebbedAtlanticblackgoby

**Comment by itsmichaeldiego** [comment]
@gpolyn That one does work! Thanks. It seems that they're changing their position a little bit, @stephenfarrar could you take a look?

**Comment by stephenfarrar** [comment]
I would guess the issue is because of rounding errors. The markers are positioned with integer coordinates, but they're also a child of a container div positioned with integer coordinates.

The container div is positioned in [google_map.js](https://github.com/google-map-react/google-map-react/blob/7877008e17427d85403a7b82e61c6ce368d93803/src/google_map.js#L650) so that it *doesn't* move when the map pans. (This requires some trickery, because everything on the map is supposed to move with the map.)

The markers are positioned in [google_map_markers.js](https://github.com/google-map-react/google-map-react/blob/7877008e17427d85403a7b82e61c6ce368d93803/src/google_map_markers.js#L272). Ordinarily nothing would need to happen when panning, but because the container doesn't move with the map, we need to do work here.

I would suggest we get rid of the containing div, or change the way it is positioned... but that's a bigger job than I have time for this week. Does that give you enough to get started?

**Comment by itsmichaeldiego** [comment]
That is a good input @stephenfarrar, unfortunately I am super busy until beginning of June, that is why I tagged you initially, I can take a look at it if no one hasn't by then though, that's for sure.

**Comment by gpolyn** [comment]
@stephenfarrar I just peeked at your linked code sections. I would like to help with the fix you propose, but would need more guidance.

Meanwhile, is the problem a math precision problem that can be solved in the short-run with [decimal.js](http://mikemcl.github.io/decimal.js/) or similar?

cc: @itsmichaeldiego

**Comment by itsmichaeldiego** [comment]
For sure @gpolyn, I will try to help you as much as I can:

When GoogleMap `v3.32` was released, they changed some stuff that affected us, you can [see more details here](https://github.com/google-map-react/google-map-react/issues/552) in #552 . 

@stephenfarrar Did a great job fixing this in #559 , and because of his fix and other previous fixes that were game-changers, I bumped the version from `0.34.0` to `1.0.0`.

So, if the jiggling started happening for you when you upgraded from `0.34.0` to `1.0.0`, PR https://github.com/google-map-react/google-map-react/pull/559 is most-likely the reason.

Now, I prefer this little jiggle instead of the problem we were having in #559, so thumbs up for @stephenfarrar 👍. Now lets work together to improve it @gpolyn 

Thanks everyone <3

**Comment by stephenfarrar** [comment]
@gpolyn Yes, you are correct that the rounding is happening inside `fromLatLngToDivPixel` and `fromLatLngToContainerPixel`. This is the expected maps API behaviour and we can't really change that.

In the marker file, I think that we need to call `fromLatLngToDivPixel` instead of `fromLatLngToContainerPixel`, and then subtract off the offset of the container. This would remove the double rounding. We would need to give the offset of the container to the geo service.

The alternative is to keep the container at (0, 0). This would be simpler for marker positioning, but probably has other implications I'm unaware of. At the moment I keep offsetting the container because it's closest to the old behaviour.

**Comment by gpolyn** [comment]
@stephenfarrar, @itsmichaeldiego - final proof that the problem presents itself in the master, that the issue is not a next.js thing: https://gfycat.com/NaughtyCalmLhasaapso

**Comment by gpolyn** [comment]
@stephenfarrar happy to lend a hand, but would hope you can give some more guidance. 

https://github.com/gpolyn/google-map-react/tree/issue_575

**Comment by stephenfarrar** [comment]
Thanks for the "final proof". I was wondering, since I haven't been able to repro it myself.

Maybe it only happens when the browser is under load?

I will definitely take a look.
It's probably related to http://issuetracker.google.com/74225068

**Comment by itsmichaeldiego** [comment]
@gpolyn @stephenfarrar I can reproduce it on my end, I don't think it has anything to do with SSR, I think @stephenfarrar was right in his comment above https://github.com/google-map-react/google-map-react/issues/575#issuecomment-391152629

**Comment by stephenfarrar** [comment]
I have been working on this, and I have a fix.

Here's the diagnosis. There are two problems.

1. The markers were inside a container element, and we were moving both the markers and the container. This caused two stages of rounding, and rounding errors, causing the jiggle. I am going to stop moving the container element. It will simply follow the map panes. This particular problem only manifested when there is subpixel map panning, e.g., on a hdpi device.

2. There is a bug in the Maps API, which will be fixed soon in v3.33. http://issuetracker.google.com/74225068 . That bug causes a different sort of jiggle at the end of a gesture or animation.

**Comment by itsmichaeldiego** [comment]
@stephenfarrar Dude you're the best! Thank you so much.

Regarding the container, I believe there is some logic of the container that is a little bit over-engineered , there are a couple of things that we can totally remove and make it way simpler. I want to work on that as soon as I've some more free time.

**Comment by itsmichaeldiego** [comment]
@stephenfarrar I've fixed the build in #603 and added just one little comment regarding naming, but the PR looks great, looking forward to merge it soon!

**Comment by itsmichaeldiego** [comment]
@stephenfarrar @gpolyn Released in 1.0.5, let me know if everything is working as expected, for me it does.

**Comment by itsmichaeldiego** [comment]
@gpolyn @stephenfarrar Using 1.0.5 works for me, what about you?

**Comment by gpolyn** [comment]
@stephenfarrar @itsmichaeldiego yes I confirm. Thank you both so much for this correction, which will allow me to continue dev'ing with the package.

**Comment by itsmichaeldiego** [comment]
@gpolyn I am glad to hear that! Thanks for reporting and helping us fix the issue.
